src/mgr/FunAction.cc

```diff
@@ -50,9 +50,7 @@ Mgr::FunAction::dump(StoreEntry* entry)
     debugs(16, 5, MYNAME);
     Must(entry != nullptr);
     if (UsingSmp())
-        storeAppendPrintf(entry, "by kid%d {\n", KidIdentifier);
+        storeAppendPrintf(entry, "---\nkid: %d\n", KidIdentifier);
```

Comment:
No, we don't.
For instance, in the updated server_list report, we do not emit ICP or HTCP stats if they are not enabled.

In all libraries I've tested, multiple documents get rendered as vectors of documents, and the size of this vector is the authority information for the number of kids. Specifying a kid number is useful because kids might be reordered in different runs but I believe we can safely skip it when there is only one. It would be useful to mark aggregate statistics as such but that's a separate topic

Reply:
> No, we don't. For instance, in the updated server_list report, we do not emit ICP or HTCP stats if they are not enabled.

You are right that the set of fields, in the sense you are interpreting that phrase, cannot be constant. I agree that we can treat kid identification as an optional field that may be absent from some reports.

> Specifying a kid number is useful because kids might be reordered in different runs

I agree that reporting kid identifier is useful (mostly for humans), even when that information is redundant. I do not think kid reports might be reordered in the current implementation. However, even if we manage to preserve the natural order forever, some kid reports may be _missing_ (because a kid has not started yet or was not revived after death).

> It would be useful to mark aggregate statistics as such but that's a separate topic

Agreed on both counts, but the two topics intersect because we may decide to use the _same_ field to report all three variations (option A):

* (no `kids` field): non-aggregated, non-SMP output
* `kids: [2]`: non-aggregated, SMP output (produced by kid2)
* `kids: [1,2,3]`: aggregated SMP output (aggregated based on kid1, kid2, and kid3 information)
* `kids: [2]`: aggregated, SMP output (aggregated based on kid2 information)

Cons: No way to distinguish reports aggregated across a single kid (yes, that may happen; see "missing" above for one example) from non-aggregated SMP reports, as illustrated by the second and last bullet using identical spelling. Since aggregation is a complicated action even when one kid is involved, this difference may be important enough to signal about.

The alternative (option B) is to use two different fields:

* (no `kid` or `kids` field): non-aggregated, non-SMP output
* `kid: 2`: non-aggregated, SMP output (produced by kid2)
* `kids: [1,2,3]`: aggregated SMP output (aggregated based on kid1, kid2, and kid3 information)
* `kids: [2]`: aggregated, SMP output (aggregated based on kid2 information)

Cons: Consumers will see a change in the report field name (from `kid` to `kids`) as we aggregate more reports (this kind of goes back to my earlier poorly worded argument about the constant set of fields being a desirable design goal).

If we pick option A, then this PR code will be affected.

My current weak preference is for option B: I think option A "cons" is (more) significant.

What do you think?

Reply:
I would put forward another option, and a proposal for a strategy:
* kids: 2 for non-aggregated SMP output
* kids: all for aggregated output
* no kids key at all for output that needs no aggregation

I don't expect there will ever be a case where we havde a partial collection of kids, therefore listing the kids is unnecessary complexity on the callers

Reply:
> I would put forward another option, and a proposal for a strategy:
 
I have added quotes to bullets to show where I think the examples start/end in your proposal.

> * "kids: 2" for non-aggregated SMP output

"kids: 2" reads like "two kids" to me which is kind of the opposite of the intended (AFAICT) "the second kid" or, more precisely, "kid identified as kid2" meaning.

> * "kids: all" for aggregated output

IMO, from machine consumer point of view, we should not change the _type_ of the value from number (e.g., "2" in non-aggregated output) to a string (i.e. "all"), especially when the reverse conversion from the string to a number is impossible. 
 
> * no kids key at all for output that needs no aggregation

I am not sure how "needs no aggregation" is defined in this context. Are you thinking about `mgr:menu` or a report from a single kid?

> I don't expect there will ever be a case where we havde a partial collection of kids

This expectation directly contradicts the "missing" kids examples in my previous [comment](https://github.com/squid-cache/squid/pull/1784#discussion_r1570814582). Please clarify whether you consider my examples wrong/impossible or unimportant (or perhaps you just did not notice them).

Another example where aggregation may work across a subset of kids are reports specific to certain kid _roles_. For example, a disk utilization report aggregated exclusively across "disker" kids (i.e. skipping "worker" kids and "coordinator").

> therefore listing the kids is imposing unnecessary complexity on the users

The "unnecessary" assertion seems false to me: Even for the typical case where reports from _all_ kids are aggregated, the list of kid IDs is often useful. Imagine, for example, that you are receiving a report from an admin saying that Squid aggregation appears to have half of the expected values. Without the list of kid identifiers, we will not be able to tell whether aggregation code is indeed broken or half of the kids have died.

Reply:
> "kids: 2" reads like "two kids" to me which is kind of the opposite of the intended (AFAICT) "the second kid" or, more precisely, "kid identified as kid2" meaning.

Right. That should be "kid".
```
kids: [ vector ]
```
aka
```
kids:
  - 2
```

Could be easier for a machine reader, while:
```
kid: <number>
```
is better for a human reading it. This is, I believe, the crux of the matter, and the reason why by this logic alone it will be hard , if possible at all, to come to an agreement.

Your argument about being helpful to know the status of worker processes is valid, but it is IMO not best positioned here; it would deserve its own report page and I'll be happy to work on that in due time.

Reply:
The top-level collection can be anonymous. The existing one for "page" already is. Please make the entry within the collection keep a named ID/number ("kid" was/is fine IMO).

```
---
kid: 1\n
...
---
kid:2\n
...
```

Reply:
> Francesco: Your argument about being helpful to know the status of worker processes is valid

That is not my argument.

My argument is that the recipient of a report aggregated across kids X and Y must know the IDs of those kids (in some cases). Without that information, it is impossible (in some cases) to interpret the report correctly. If we were able to guarantee that all aggregated reports are aggregated across all kids at all times, then there would be some wiggle room here to omit that information, but the reality we have to deal with is more complex (as detailed already), and our reporting format/approach has to reflect that complexity.

> Francesco: status of worker processes ... is IMO not best positioned here; it would deserve its own report page and I'll be happy to work on that in due time.

Sure, but a future "status of kids" report is not my concern, and I am not requesting or even suggesting any such work here.

> Amos: The top-level collection can be anonymous.

FWIW, I do not know what "top-level collection" is in this context. If it means "report aggregated across one or more kids", then it should disclose what those kids were (as detailed earlier), either in all cases or in cases where aggregation does not include all kids.

FWIW, I recommend getting back to _comprehensive_ proposals (e.g., option A and option B in https://github.com/squid-cache/squid/pull/1784#discussion_r1570814582 and another option at https://github.com/squid-cache/squid/pull/1784#discussion_r1574292773) that cover all known use cases. Each individual use case is trivial to handle. It is the _combination_ that requires some effort.

Reply:
I disagree with you on one count: there are only two valid states here, one is "we are not aggregating data" and the other is "we are aggregating data from all workers that are responding". Piggybacking the information that "only some workers are responding" here is misplacing that kind of information - that status information belongs to some other place, such as the "info" page or a new dedicated report page if enough information needs to be shown.

Furthermore, we don't really have the infrastructure in place to accumulate which workers we are aggregating information from. It can be added, but then we fall victims of scope creep and as a result nothing will get done as it happened other times.
My bet is that by insitsting to emit a "kids: [ vector ] " information, all that will ever be output is "kids: [ exactly one entry in the vector ]", because I have no interest in going further than that, and others who might have interest have no time